cardano-crypto-class/src/Cardano/Crypto/PackedBytes.hs

```diff
@@ -0,0 +1,368 @@
+indexWord64BE (ByteArray ba#) (I# i#) =
+#ifdef WORDS_BIGENDIAN
+  W64# (indexWord8ArrayAsWord64# ba# i#)
+#else
+  W64# (byteSwap64# (indexWord8ArrayAsWord64# ba# i#))
+#endif
+{-# INLINE indexWord64BE #-}
+
+peekWord64BE :: Ptr a -> Int -> IO Word64
+peekWord64BE ptr i =
+#ifndef WORDS_BIGENDIAN
+  byteSwap64 <$>
+#endif
+  peekByteOff (castPtr ptr) i
+{-# INLINE peekWord64BE #-}
+
+
+writeWord64BE :: MutableByteArray s -> Int -> Word64 -> ST s ()
+writeWord64BE (MutableByteArray mba#) (I# i#) (W64# w#) =
+  primitive_ (writeWord8ArrayAsWord64# mba# i# wbe#)
+  where
+#ifdef WORDS_BIGENDIAN
+    !wbe# = w#
+#else
+    !wbe# = byteSwap64# w#
+#endif
+{-# INLINE writeWord64BE #-}
+
+#elif WORD_SIZE_IN_BITS == 32
+
+indexWord64BE :: ByteArray -> Int -> Word64
+indexWord64BE ba i =
+  (fromIntegral (indexWord32BE ba i) `shiftL` 32) .|. fromIntegral (indexWord32BE ba (i + 4))
+{-# INLINE indexWord64BE #-}
+
+peekWord64BE :: Ptr a -> Int -> IO Word64
+peekWord64BE ptr i = do
+  u <- peekWord32BE ptr i
+  l <- peekWord32BE ptr (i + 4)
+  pure ((fromIntegral u `shiftL` 32) .|. fromIntegral l)
+{-# INLINE peekWord64BE #-}
+
+writeWord64BE :: MutableByteArray s -> Int -> Word64 -> ST s ()
+writeWord64BE mba i w64 = do
+  writeWord32BE mba i (fromIntegral (w64 `shiftR` 32))
+  writeWord32BE mba (i + 4) (fromIntegral w64)
+{-# INLINE writeWord64BE #-}
+
+#else
+#error "Unsupported architecture"
+#endif
+
+
+indexWord32BE :: ByteArray -> Int -> Word32
+indexWord32BE (ByteArray ba#) (I# i#) =
+#ifdef WORDS_BIGENDIAN
+  W32# (indexWord8ArrayAsWord32# ba# i#)
+#else
+  W32# (narrow32Word# (byteSwap32# (indexWord8ArrayAsWord32# ba# i#)))
+#endif
+{-# INLINE indexWord32BE #-}
+
+peekWord32BE :: Ptr a -> Int -> IO Word32
+peekWord32BE ptr i =
+#ifndef WORDS_BIGENDIAN
+  byteSwap32 <$>
+#endif
+  peekByteOff (castPtr ptr) i
+{-# INLINE peekWord32BE #-}
+
+
+writeWord32BE :: MutableByteArray s -> Int -> Word32 -> ST s ()
+writeWord32BE (MutableByteArray mba#) (I# i#) (W32# w#) =
+  primitive_ (writeWord8ArrayAsWord32# mba# i# wbe#)
+  where
+#ifdef WORDS_BIGENDIAN
+    !wbe# = w#
+#else
+    !wbe# = narrow32Word# (byteSwap32# w#)
+#endif
+{-# INLINE writeWord32BE #-}
+
+byteArrayToShortByteString :: ByteArray -> ShortByteString
+byteArrayToShortByteString (ByteArray ba#) = SBS ba#
+{-# INLINE byteArrayToShortByteString #-}
+
+byteArrayToByteString :: ByteArray -> ByteString
+byteArrayToByteString ba
+  | isByteArrayPinned ba =
+    BS.fromForeignPtr (pinnedByteArrayToForeignPtr ba) 0 (sizeofByteArray ba)
+  | otherwise = SBS.fromShort (byteArrayToShortByteString ba)
+{-# INLINE byteArrayToByteString #-}
+
+pinnedByteArrayToForeignPtr :: ByteArray -> ForeignPtr a
+pinnedByteArrayToForeignPtr (ByteArray ba#) =
+  ForeignPtr (byteArrayContents# ba#) (PlainPtr (unsafeCoerce# ba#))
+{-# INLINE pinnedByteArrayToForeignPtr #-}
+
+
+withReadPtrByteString :: ByteString -> (Ptr b -> IO a) -> a
+withReadPtrByteString bs f =
+  accursedUnutterablePerformIO $
```

Comment:
You bring up a good point, this function is named so terrifyingly that it makes everyone run for the hills. 

Changed to a better named `unsafeInlineIO`

> Is that really worth the performance difference? 

yes

> This is here for performance reasons, right?

Yes, and also because it is safe to use in this use case.

Here is a good rule of thumb for you in the future, in case you ever wonder if it is safe to use `unsafeInlineIO` or not: whenever a `Ptr` or `MutableByteArray` are guaranteed to never be mutated and you are trying to use them to only read in that region of memory then you are totally safe to use, and performance gains can be massive.

Reply:
The scary-named function was preferable.

```
Generally, do not use this function. It is the same as accursedUnutterablePerformIO from bytestring and is well behaved under narrow conditions. See the documentation of that function to get an idea of when this is sound. In most cases GHC.IO.Unsafe.unsafeDupablePerformIO should be preferred.
```

This suggests that the conditions for use are documented in `bytestring`, but seems to no longer be the case. It was replaced with a history of bug tickets.

If we realllly need this (which is worth actually demonstrating), it deserves an explanation alongside it and a review from someone who understands it.

Reply:
[This bug](https://gitlab.haskell.org/ghc/ghc/-/issues/3486) in particular arose from inlining `memchr`, which only reads memory.

Reply:
@redxaxder I thought I explained it why this use case is safe, was it not clear? :wink: I am being sarcastic of course, this is all indeed a pretty complicated and dangerous stuff, so I would highly recommend you to do a bit of research into why and when this function(s) can and cannot be used, this is really fun stuff. There are plenty discussions online on  this topic,  I am not going to create a full explanation in this codebase as a lengthy comment, cause that would have to be a blog post length comment.

> is well behaved under narrow conditions.

This is precisely this set of conditions. To give you some intuition, here is an example usage from `bytestring`:

https://github.com/haskell/bytestring/blob/36b9fde3249b0645213d20c28b0513d06edea02a/Data/ByteString/Unsafe.hs#L105-L108

i.e. indexing into a `ByteString`, which is exactly what we are doing here, no more no less. I hope this gives you a bit more confidence. 

> which is worth actually demonstrating

Be my guest, feel free to add some benchmarks and compare the performance. (I've done it plenty of times already to convince myself.)

Reply:
In case of `memchr` in https://gitlab.haskell.org/ghc/ghc/-/issues/3486 it was due to a lazy thunk being returned

Reply:
In fact even `unsafePerformIO` would not make that problem go away in https://gitlab.haskell.org/ghc/ghc/-/issues/3486, it was incorrect usage of `touch`, which we also recently had in this repo.

Also `elemIndices` still uses `accursedUnutterablePerformIO`: and `memchr`  https://github.com/haskell/bytestring/blob/9dee131929f6a6e65c815f384557995fe098e154/Data/ByteString.hs#L1289-L1298

Reply:
I added a short comment about usage of `unsafeInlineIO`

Reply:
> I am not going to create a full explanation in this codebase as a lengthy comment, cause that would have to be a blog post length comment.

A link to an explanation is also fine. We want this code to be editable by people in the future, many of which will not have prior familiarity with this thing.

I asked Duncan about it and I'm convinced that this won't misbehave with current use. He did point out that the function `withReadPtrByteString` is not safe. (Suggested name: `unsafeWithByteStringPtr`)

This lines up with the `memchr` bug. It can't be called with a function that lazily reads the value.

Reply:
> Be my guest, feel free to add some benchmarks and compare the performance. (I've done it plenty of times already to convince myself.)

I don't doubt that the read itself is faster.

The justification for use rests on whether this impacts a performance bottleneck our system actually has. If it doesn't, the benefit of the extra speed goes away, and we're only left with the cost of more difficult code.

Reply:
Renamed.

> A link to an explanation is also fine. 

I have a vague recollection of seeing a good summary about those unsafe functions, but I can't find it now. So, I don't have any posts that go into details when it is safe to use `unsafeInlineIO`, all posts that I can find now go into detail why it is dangerous and what sort of ugly bugs you get when it is used incorrectly. I only know this specifics due reading and writing a lot of code that deals with this stuff. Maybe I should write a blog post about it do I can link to it in the future, hopefully one day.

People who do  not have prior familiarity with this thing, should not be editing this module. This is not your regular safe day-to-day Haskell, which is sort of expected from a performance critical library that deals with cryptography.

Reply:
> I have a vague recollection of seeing a good summary about those unsafe functions, but I can't find it now.

Here's one: https://free.cofree.io/2020/07/20/perform-io/

> People who do not have prior familiarity with this thing, should not be editing this module. This is not your regular safe day-to-day Haskell, which is sort of expected from a performance critical library that deals with cryptography.

I agree that in its current form, this is not a module we would expect many people to be able to edit. Or audit.

I suspect our disagreement may be about whether it is desirable to for code to have those properties. I think they are very desirable, and a choice to throw them away shouldn't be made lightly.

Reply:
> The justification for use rests on whether this impacts a performance bottleneck our system actually has. 

How do you propose we measure the impact of using this function or another in a perspective of a whole system? In order for it to be measurable it has to be monumentally slower or faster. In this kind of scenarios only microbenchmarks can be useful.

> benefit of the extra speed goes away

I highly disagree with this statement. Aggregation of small optimizations together can have a significant impact on a performance of a system as a whole. We should not discard improvements due to ignorance.

> I suspect our disagreement may be about whether it is desirable to for code to have those properties.

I wish we had a library that could help us avoid unsafe code like that, but we don't have one yet.

> https://free.cofree.io/2020/07/20/perform-io/

I did skim the post, but I don't see a mention anywhere of use cases when `accursedUnutterablePerformIO` is safe to use.

The only place I found today that even mentions remotely it's potential safe usage is this SO: https://stackoverflow.com/questions/61021205/what-is-the-difference-between-unsafedupableperformio-and-accursedunutterableper : 

> When accursedUnutterablePerformIO works correctly, it can potentially give slightly better performance because there are more opportunities for optimization if the m realWorld# call can be inlined earlier rather than later. So, the actual bytestring library still uses accursedUnutterablePerformIO internally in lots of places, in particular where there's no allocation going on (e.g., head uses it to peek the first byte of the buffer).

Reply:
> How do you propose we measure the impact of using this function or another in a perspective of a whole system?

The part of our system that seems most likely to be impacted by this is block deserialization.

>> benefit of the extra speed goes away
> I highly disagree with this statement.

As an extreme example, you wouldn't rewrite a module in C to gain a microsecond improvement on an hour long task. (I don't think that this situation is analogous to that one. Just illustrating a principle.)

Do we actually disagree on the underlying principle?

Or just on where this particular approach sits in relation to it?

> I thought I explained [...] was it not clear?
> [...] due to ignorance.

This kind stuff is distracting from the substance of the conversation.

Reply:
I agree that you want to use `accursedUnutterablePerformIO` and not `unsafeInlineIO` because the whole point is to highlight the proof obligations. Using a less-scary name just hides that. That was the whole reason that the function in the `bytestring` package was renamed to the long scary `accursedUnutterablePerformIO`.

Reply:
>  you want to use accursedUnutterablePerformIO and not unsafeInlineIO

Done.